proposals/shared-everything-threads/Overview.md

```diff
@@ -194,37 +107,260 @@ how a non-shared function accesses a `shared` memory):
+conclusions are `ok(shared)` iff the premises are `ok(shared)`. The validation rules for
+`absheaptype` and `comptype` are extended so that they are only `ok(shared)` if the heap type is
+shared. Number types and vector types are always valid as shared.
+
+Similarly, we need to be able to validate that a `typeidx` is valid as shared. We introduce an
+auxiliary function `expandshare` similar to `expand` that returns `shared` or `unshared` for a given
+deftype. A `typeidx` `x` is `ok(shared)` iff `expandshare(C.types[x]) = shared`.
+
+Shared and unshared types are never subtypes of one another. The subtype relationships among shared
+abstract types mirrors the subtype relationships among unshared abstract types.
+
+> Note: Should i31ref behave like an i32 in that it is usable in both shared and non-shared contexts
+> without needing to have a separate `shared` variant? This seems like it would be ok, except we
+> would still need a shared version to interoperate with `(shared eq)`
+
+#### Element Segments
+
+The syntax of `elem` is extended:
+
+```
+elem ::= {type reftype, init vec(expr), mode elemmode, share}
+```
+
+An element segment is valid as shared only if its reftype is valid as shared. Instructions that
+refer to element segments (e.g. `elem.drop`) are only valid as shared if their element segment is
+shared.
+
+#### Data Segments
+
+The syntax of data segments is extended:
+
+```
+data ::= {init vec(byte), mode datamode, share}
+```
 
-| Section       | Proposed change (rough)                                                                  |
-|---------------|------------------------------------------------------------------------------------------|
-| Syntax        | `thread.hw_concurrency`                                                                  |
-| Validation    | Valid with type `[] -> [i32]`                                                            |
-| Execution     | Push on the stack the number of threads the module _may_ be able to execute concurrently |
-| Binary format | `0xFE 0x05 => thread.hw_concurrency`                                                     |
+Instructions that refer to data segments (e.g. `data.drop`) are only valid as shared if their data
+segment is shared.
 
-This instruction has fingerprinting potential, which is a concern in some environments.
+> Note: we could alternatively make all data segments shared, just like we do for other
+> non-reference data.
+
+#### Exception tags
+
+The syntax of tags does not need to be extended. They are shared iff their function types are
+shared. Shared tags are used to throw and catch `(shared exnref)`. Instructions that reference tags
+(e.g. `throw` and `catch`) only validate as shared if all the tags they reference are shared.
+
+> TODO: We need to determine how sharing affects exception handling. We may want to add
+> `catch_all_shared_ref`. We may want `(shared exnref) <: exnref` so `catch_all_ref` can continue to
```

Comment:
Yes, I agree it would be nice to be consistent. But it would also be nice to have a reasonable story for how EH interacts with sharedness. We'll have to explore the options more.